[V1][Metrics] Add GPU prefix cache hit rate % gauge

comaniac opened the pull request:
Part of https://github.com/vllm-project/vllm/issues/10582

cc @markmc @robertgshaw2-redhat

comaniac committed:
done

Signed-off-by: Cody Yu <hao.yu.cody@gmail.com>

markmc commented on vllm/v1/core/kv_cache_manager.py:
```diff
@@ -109,6 +132,9 @@ def get_computed_blocks(
                 break
 
+        self.prefix_caching_metrics["query_total"] += len(block_hashes)
+        self.prefix_caching_metrics["query_hit"] += len(computed_blocks)
```
I see you decided to drop the great care you went to in v0 to avoid large integer values - https://github.com/vllm-project/vllm/pull/7606#discussion_r1720372929

Seems reasonable :+1:

comaniac replied:
Yeah decided to keep it simple first and revisit later if needed.

markmc commented on vllm/v1/metrics/loggers.py:
```diff
@@ -104,6 +106,12 @@ def __init__(self, model_config: ModelConfig):
             documentation="GPU KV-cache usage. 1 means 100 percent usage.",
             labelnames=labelnames).labels(*labelvalues)
 
+        self.gauge_gpu_prefix_cache_hit_rate = prometheus_client.Gauge(
```
I know this was a Gauge in v0 too and every other metric we've added to v1, we've just re-used the v0 metric

But, in this case, I think it's worth considering adding `query_total` and `query_hit` Counters instead, and using Prometheus queries like this instead:

```rate(cache_query_hit[5m]) / rate(cache_query_total[5m])```

Consider how e.g. the hit rate drops due to a sudden increase in misses, that change will be diluted by the large accumulated totals.

You might be monitoring one set of instances that have just started up, and another set of instances that have been running for weeks ... if the hit rate drops consistently across all of those instances, it would only show up in the metrics of the recently started instances.

markmc replied:
Actually, the right thing to do would be to add the `hit_rate` Gauge with a "deprecated" note in the documentation string, and also add the two Counters :+1:

comaniac replied:
That makes sense, but we should align the time interval (between two logs) of all metrics, so maybe we should just reset the metrics in the kv cache manager after `scheduler_stats.gpu_prefix_cache_hit_rate` is called. In this case we also don't have to deprecate `hit_rate`.

markmc replied:
With Prometheus, the time interval for the rate should be a user-defined thing - https://prometheus.io/docs/prometheus/latest/querying/functions/#rate - that's why you'd want the counters

With logging, yep I agree it would be good to reset the counters and calculate the hit rate over the interval between log outputs

I'm happy to take on figure this out if you like :+1:

markmc replied:
I think if we want to have the hit rate in LoggingStatLogger, we would need it to accumulate and reset its own counters

comaniac replied:
Yeah that makes sense. If possible I still want to reset metrics in the scheduler so that the large number issue can be avoided at all. I'll try to figure out

markmc reviewed (commented):
Thanks Cody!

comaniac committed:
reset

Signed-off-by: Cody Yu <hao.yu.cody@gmail.com>

markmc commented on vllm/v1/core/scheduler.py:
```diff
@@ -545,6 +545,8 @@ def make_stats(self) -> SchedulerStats:
             num_waiting_reqs=len(self.waiting),
             gpu_cache_usage=self.kv_cache_manager.usage,
+            gpu_prefix_cache_hit_rate=self.kv_cache_manager.
+            get_and_reset_prefix_cache_hit_rate(),
```
This resets the hit rate on every scheduler step?

comaniac committed:
improve

Signed-off-by: Cody Yu <hao.yu.cody@gmail.com>

comaniac committed:
test

Signed-off-by: Cody Yu <hao.yu.cody@gmail.com>

comaniac commented:
After the second thought. I feel it might be easier to simply maintain the cache hit rate of the most recent N (e.g., 1k) requests. This aligns the logging flow better because we take a snapchat of the scheduler stat in every step. It also makes more sense to me because the cache hit rate doesn't associate with time but only requests. WDYT?

comaniac commented:
Gentle ping @markmc

markmc committed:
[V1][Metrics] Replace prefix cache hit rate with queries/hits counters

With a hit rate metric, we need to avoid the situation where if the hit
rate drops due to a sudden increase in misses, the change gets diluted
by the large accumulated totals.

With Prometheus, it is better to log monotonic counters and use
Prometheus queries to calculate rates according to whatever time
interval makes sense to the operator, also allowing them to easily
adjust the time interval in their PromQL queries.

With the logging logger, we can continue to aggregate across an
interval of (roughly) 1000 requests.

Signed-off-by: Mark McLoughlin <markmc@redhat.com>

markmc commented:
> Gentle ping @markmc

Sorry for taking so long to come back to this!

I took a look again and played with it, and I'm pretty convinced:

1. The right thing to do in Prometheus is to log the queries and hits as a counter
2. The logging logger should calculate the hit rate over the logging interval
3. We probably don't need to worry about retaining the v0 hit rate gauge - e.g. it's not in the default Grafana dashboard. If any users complain, maybe we can add it back to v1 for a deprecation period.

See https://github.com/markmc/vllm/commit/4227ffcd26b8ab1bc432594b09eb0d4380d2ef01 - I've implemented this approach on top of your changes, and explained in more detail in the commit message. HTH! :+1:

comaniac commented:
Sorry unfortunately I'm not convinced...

1. What about the potential overflow issue of query and hit if an engine is served for months?
2. To me using time interval (e.g., 5 seconds) can probably satisfy what you want to observe. However, 1k requests can achieve the same goal, too. The key to achieve it is not about using number of requests or time to form an interval. It's about whether the interval is configurable.

For example I want to benchmark the overall hit rate of a dataset with 2k requests, I can configure the metric to set the interval to be "2k requests" or "1 day" (i.e., infinity) to obtain the result I want. On the other hand, if it's not configurable, then both methods cannot work in this case. Thus, I'd prefer the method that is potentially more sustainable.

markmc commented:
> Sorry unfortunately I'm not convinced...
> 
> 1. What about the potential overflow issue of query and hit if an engine is served for months?

With [my commit](https://github.com/markmc/vllm/commit/4227ffcd26b8ab1bc432594b09eb0d4380d2ef01), there is no longer any accumulation of these values in engine core.

In the case of Prometheus, the client is storing this as a float and `sys.float_info.max = 1.7976931348623157e+308`. But also ... this is exactly what a Prometheus counter is designed for.

In the case of logging, the accumulators are reset every logging interval.

> 2. To me using time interval (e.g., 5 seconds) can probably satisfy what you want to observe. However, 1k requests can achieve the same goal, too. The key to achieve it is not about using number of requests or time to form an interval. It's about whether the interval is configurable.
> 
> For example I want to benchmark the overall hit rate of a dataset with 2k requests, I can configure the metric to set the interval to be "2k requests" or "1 day" (i.e., infinity) to obtain the result I want. On the other hand, if it's not configurable, then both methods cannot work in this case. Thus, I'd prefer the method that is potentially more sustainable.

I think it would make sense to make the logging interval configurable (from 5s), but we would do well to avoid adding configurable intervals for different logged metrics - i.e. VLLM_LOG_STATS_INTERVAL seems fine to me, but a proliferation of more stuff like VLLM_LOG_BATCHSIZE_INTERVAL doesn't seem like such a good direction.

We could also have the logger print out the actual `cache_queries` and `cache_hits` values for you to gather and calculate rates over whatever buckets you want ... which actually sounds closer to your use case.

You could also consider using Prometheus to gather metrics during benchmarking.

comaniac commented:
> > Sorry unfortunately I'm not convinced...
> > 
> > 1. What about the potential overflow issue of query and hit if an engine is served for months?
> 
> With [my commit](https://github.com/markmc/vllm/commit/4227ffcd26b8ab1bc432594b09eb0d4380d2ef01), there is no longer any accumulation of these values in engine core.
> 
> In the case of Prometheus, the client is storing this as a float and `sys.float_info.max = 1.7976931348623157e+308`. But also ... this is exactly what a Prometheus counter is designed for.

Yeah I know your commit accumulates values in the logger (both the LoggingStatLogger and Prometheus, where the LoggingStatLogger stores in integer). However, LoggingStatLogger still reports the overall cache hit rate rather than an interval IIUC?

> 
> In the case of logging, the accumulators are reset every logging interval.
> 
> I think it would make sense to make the logging interval configurable (from 5s), but we would do well to avoid adding configurable intervals for different logged metrics - i.e. VLLM_LOG_STATS_INTERVAL seems fine to me, but a proliferation of more stuff like VLLM_LOG_BATCHSIZE_INTERVAL doesn't seem like such a good direction.
> 

I agree that we should avoid adding configurable intervals for different logged metrics, and that's why I don't feel resetting the accumulator every logging interval is feasible. For example what if I still want to see the log every 5 seconds but prefer to get the hit rate of a larger window?

> We could also have the logger print out the actual `cache_queries` and `cache_hits` values for you to gather and calculate rates over whatever buckets you want ... which actually sounds closer to your use case.

This approach would introduce two issues IMO:
1. The logging may be messing up because of large numbers.
2. Users would need to parse the log and calculate the rates in a separate process.

> You could also consider using Prometheus to gather metrics during benchmarking.

But we do support LoggingStatLogger and PrometheusStatLogger, and LoggingStatLogger is more straightforward and lightweight to many cases. Unless we make a clear document about which logger should be used in which case and what's differences in their metrics, we should do our best to make sure both loggers behave consistent.

walker-ai commented:
Does prefix caching compatible with lora adapters? I found that if I enable lora and prefix_caching it will throw assertion error.

comaniac commented:
> @comaniac check out [markmc@9b24079](https://github.com/markmc/vllm/commit/9b240796501c543eedfde88e498f058544a98ec2) - it retains your approach for the logger, but exposes the counters for Prometheus

In general LGTM. Thanks. It's just a bit weird for PrefixCacheStats and PrefixCachingMetrics to be located in different places, but I guess this is the best way for now. If you want you can just commit to my branch and I'll make some minor changes. Afterwards it should be good to go.

comaniac committed:
minor

Signed-off-by: Cody Yu <hao.yu.cody@gmail.com>

iamyourblackjack commented on vllm/v1/core/kv_cache_utils.py:
```diff
@@ -28,6 +30,68 @@ class BlockHashType(NamedTuple):
+        are being scheduled and are looking for computed blocks.
+
+        When there are more than `interval` requests, the oldest set of
+        requestsare removed from the metrics.
```
```suggestion
        request sare removed from the metrics.
```

ywang96 commented on vllm/v1/metrics/stats.py:
```diff
@@ -1,14 +1,27 @@
+    reset: bool = False
+    # The number of requests in this update.
+    requests: int = 0
+    # The number of queries in these requests.
+    queries: int = 0
```
I know you have added this in the prom metrics doc, It's probably also worth expanding here to explain that "queries" means number of queried blocks, otherwise it might seem a bit unclear what this means to developers without the context.

ywang96 commented on vllm/v1/metrics/loggers.py:
```diff
@@ -72,13 +78,15 @@ def log(self, scheduler_stats: SchedulerStats,
-            "GPU KV cache usage: %.1f%%.",
+            "Running: %d reqs, Waiting: %d reqs, "
+            "GPU KV cache usage: %.1f%%, "
+            "Prefix cache hit rate: %.1f%%",
             prompt_throughput,
             generation_throughput,
             scheduler_stats.num_running_reqs,
             scheduler_stats.num_waiting_reqs,
             scheduler_stats.gpu_cache_usage * 100,
+            self.prefix_caching_metrics.hit_rate * 100,
```
Hmm.. I wonder if it's worth the effort to have a separate `logger.info` (that depends on if prefix caching is enabled)? 

If I understand correctly, this will always be 0 if prefix cachng is turned off, right? (In V0, it's always -1 in this case, so we might have to keep the same behavior)

comaniac replied:
Good point. In fact in v0 I disabled this log when prefix caching is disabled, but given that prefix caching is a first class citizen I feel it might be fine. Open to other opinions tho

ywang96 replied:
Sounds good!

ywang96 reviewed (approved):
Overall makes sense to me and I left two comments! PTAL and resolve the merge conflict before merging this PR. Thanks!

comaniac committed:
comment

Signed-off-by: Cody Yu <hao.yu.cody@gmail.com>

iamyourblackjack reviewed (commented):
Thanks for working on this. It would be nice if we can have it for the next release.
